julia-repl-comint.el

```diff
@@ -0,0 +1,276 @@
+
+(defun julia-repl-comint-question-mark ()
+  "Switch to the help interface at the Julia prompt or insert '?'."
+  (interactive)
+  (when (looking-back (concat julia-repl-julia-prompt-regexp "\\(;\\|\\]\\)") nil)
+    (julia-repl-comint-backspace))
+  (insert "?")
+  (when (looking-back (concat julia-repl-julia-prompt-regexp "\\?"))
+    (let ((prompt "help?> ")
+	  (bol (save-excursion (search-backward-regexp "^") (point))))
+      (when julia-repl-comint-use-prompt-faces
+       (put-text-property 0 (1- (length prompt)) 'face 'julia-repl-help-prompt-face prompt))
+      (put-text-property bol (point) 'display prompt))))
+
+(defun julia-repl-comint-semicolon ()
+  "Switch to the shell interface at the Julia prompt or insert ';'."
+  (interactive)
+  (when (looking-back (concat julia-repl-julia-prompt-regexp "\\(\\?\\|\\]\\)") nil)
+    (julia-repl-comint-backspace))
+  (insert ";")
+  (when (looking-back (concat julia-repl-julia-prompt-regexp ";"))
+    (let ((prompt "shell> ")
+	  (bol (save-excursion (search-backward-regexp "^") (point))))
+      (when julia-repl-comint-use-prompt-faces
+       (put-text-property 0 (1- (length prompt)) 'face 'julia-repl-shell-prompt-face prompt))
+      (put-text-property bol (point) 'display prompt))))
+
+(defun julia-repl-comint-backspace ()
+  (interactive)
+  (let ((bol (save-excursion (search-backward-regexp "^") (point))))
+    (cond
+     ((looking-back julia-repl-julia-prompt-regexp)
+      nil)
+     ((looking-back (concat julia-repl-julia-prompt-regexp "."))
+      (put-text-property bol (point) 'display nil)
+      (delete-backward-char 1))
+     (t
+      (delete-backward-char 1)))))
+
+
+(defun julia-repl-comint-maybe-send-input ()
+  "Send the input to the julia process (or add a newline and indent if
+it's an incomplete julia expression)."
+  (interactive)
+  (when completion-in-region-mode
+    (completion-in-region-mode -1))
+  (let ((proc (get-buffer-process (current-buffer))))
+    (if (not proc) (user-error "No Julia process is running!")
+      (widen)
+      (let* ((pmark (process-mark proc))
+             (intxt (if (>= (point) (marker-position pmark))
+                        (progn (if comint-eol-on-send
+				   (if comint-use-prompt-regexp
+				       (end-of-line)
+				     (goto-char (field-end))))
+                               (buffer-substring pmark (point)))
+                      (let ((copy (funcall comint-get-old-input)))
+                        (goto-char pmark)
+                        (insert copy)
+                        copy)))
+             (input (if (not (eq comint-input-autoexpand 'input))
+                        ;; Just whatever's already there.
+                        intxt
+                      ;; Expand and leave it visible in buffer.
+                      (comint-replace-by-expanded-history t pmark)
+                      (buffer-substring pmark (point))))
+	     (jl-call (format "Base.parse_input_line(\"%s\")" input))
```

Comment:
A quick note, since this is command has a lot of complicated-looking elisp: everything above this line with `jl-call` is copied exactly out of `comint-send-input`. In prose, this command's purpose is: "do exactly what `comint-send-input` would normally do, except if it's incomplete julia code at a `julia> ` prompt. If so, just newline and indent."

Reply:
Thanks for the explanation.